numpy/core/tests/test_multiarray.py

```diff
@@ -7854,6 +7854,37 @@ def test_null_inside_bstring_array_is_truthy(self):
         assert_(a)
 
 
+class TestUnicodeEncoding:
+    """
+    Tests for encoding related bugs, such as UCS2 vs UCS4, round-tripping
+    issues, etc
+    """
+    @pytest.mark.xfail(IS_PYPY, reason=
+        "PyPy overloads this, and uses `.decode('utf-32-le')` without "
+        "`errors='surrogatepass'`")
```

Comment:
https://bitbucket.org/pypy/pypy/issues/3165

Reply:
Is CPython allowing you to create a unicode character `c` where `ord(c) >= sys.maxunicode`?

Reply:
Good point, I should include a test for that.

Reply:
With this patch:
```
>>> x = np.array(bytes([255]*4)).view('U1')
>>> x[()]
SystemError: invalid maximum character passed to PyUnicode_New
```
Before this patch:
```
>>> x = np.array(bytes([255]*4)).view('U1')
>>> x[()]
UnicodeDecodeError: 'utf-32-le' codec can't decode bytes in position 0-3: code point not in range(0x110000)
```

Reply:
We could revert back to using `PyUnicode_DecodeUTF32(..., "surrogatepass")`, but that would be a lot slower, for very little value.

Reply:
Sounds strange enough that I would not worry (I did follow it too closely). If only the error type changes (even if SystemError is a bit annoying), I would not worry. If there are more corner case changes along strange things like surrogates, lets just mention it in the release notes.

Reply:
Yeah, it's only the error type that changes. We could catch the exception and replace it if we wanted - or perhaps just change `PyUnicode_FromKindAndData` to raise `UnicodeDecodeError` if it discovers a codepoint that is too large.

Reply:
> This fails on PyPy

Should now be fixed since the PyPy CI run uses the latest nightly, and the bug was fixed there.

Reply:
I removed the commit that added the xfail, let's see...

Reply:
That did the trick, thanks!